scripts/research/b2_04_moderate_pullback_structure_lib.py

```
@@ -0,0 +1,1327 @@
+            raise B204Error("authorized partition evidence is malformed")
+        partition_years.append(_partition_year(str(partition["relative_path"])))
+    if not partition_years:
+        raise B204Error("authorized partition evidence is empty")
+
+    open_ms = np.asarray(frame["open_time_ms"], dtype=np.int64)
+    avail_ms = np.asarray(frame["available_at_ms"], dtype=np.int64)
+    if len(open_ms) == 0:
+        raise B204Error("loaded development frame is empty")
+
+    if end_exclusive_ms > VALIDATION_2025_START_MS:
+        raise B204Error("authorized window reaches the reserved 2025 validation pool")
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)